luminoth/tools/checkpoint/__init__.py

```diff
@@ -0,0 +1,723 @@
+    with open(path, 'w') as f:
+        json.dump(checkpoints, f)
+
+
+def merge_index(local_index, remote_index):
+    """Merge the `remote_index` into `local_index`.
+
+    The merging process is only applied over the checkpoints in `local_index`
+    marked as ``remote``.
+    """
+
+    non_remotes_in_local = [
+        c for c in local_index['checkpoints']
+        if c['source'] != 'remote'
+    ]
+    remotes_in_local = {
+        c['id']: c for c in local_index['checkpoints']
+        if c['source'] == 'remote'
+    }
+
+    to_add = []
+    seen_ids = set()
+    for checkpoint in remote_index['checkpoints']:
+        seen_ids.add(checkpoint['id'])
+        local = remotes_in_local.get(checkpoint['id'])
+        if local:
+            # Checkpoint is in local index. Overwrite all the fields.
+            local.update(**checkpoint)
+        elif not local:
+            # Checkpoint not found, it's an addition. Transform into our schema
+            # before appending to `to_add`. (The remote index schema is exactly
+            # the same except for the ``source`` and ``status`` keys.)
+            checkpoint['source'] = 'remote'
+            checkpoint['status'] = 'NOT_DOWNLOADED'
+            to_add.append(checkpoint)
+
+    # Out of the removed checkpoints, only keep those with status
+    # ``DOWNLOADED`` and turn them into local checkpoints.
+    missing_ids = set(remotes_in_local.keys()) - seen_ids
+    already_downloaded = [
+        c for c in remotes_in_local.values()
+        if c['id'] in missing_ids and c['status'] == 'DOWNLOADED'
+    ]
+    for checkpoint in already_downloaded:
+        checkpoint['status'] = 'LOCAL'
+        checkpoint['source'] = 'local'
+
+    new_remotes = [
+        c for c in remotes_in_local.values()
+        if not c['id'] in missing_ids  # Checkpoints to remove.
+    ] + to_add + already_downloaded
+
+    if len(to_add):
+        click.echo('{} new remote checkpoints added.'.format(len(to_add)))
+    if len(missing_ids):
+        if len(already_downloaded):
+            click.echo('{} remote checkpoints turned to local.'.format(
+                len(already_downloaded)
+            ))
+        click.echo('{} remote checkpoints removed.'.format(
+            len(missing_ids) - len(already_downloaded)
+        ))
+    if not len(to_add) and not len(missing_ids):
+        click.echo('No changes in remote index.')
+
+    local_index['checkpoints'] = non_remotes_in_local + new_remotes
+
+    return local_index
+
+
+def get_checkpoint(db, id_or_alias):
+    """Returns checkpoint entry in `db` indicated by `id_or_alias`.
+
+    First tries to match an ID, then an alias. For the case of repeated
+    aliases, will first match local checkpoints and then remotes. In both
+    cases, matching will be newest first.
+    """
+    # Go through the checkpoints ordered by creation date. There sholdn't be
+    # repeated aliases, but if there are, prioritize the newest one.
+    locals = sorted(
```

Comment:
Maybe locals is a bad var name? https://docs.python.org/2/library/functions.html#locals